grpc/src/main/java/io/stargate/grpc/service/QueryHandler.java

```diff
@@ -0,0 +1,258 @@
+            });
+  }
+
+  private CompletionStage<Result> executePrepared(Prepared prepared) {
+    long queryStartNanoTime = System.nanoTime();
+
+    Payload values = query.getValues();
+    PayloadHandler handler = PayloadHandlers.get(values.getType());
+
+    QueryParameters parameters = query.getParameters();
+
+    try {
+      return connection.execute(
+          bindValues(handler, prepared, values), makeParameters(parameters), queryStartNanoTime);
+    } catch (Exception e) {
+      CompletableFuture<Result> failedFuture = new CompletableFuture<>();
+      failedFuture.completeExceptionally(e);
+      return failedFuture;
+    }
+  }
+
+  private CompletionStage<Result> reprepareAndRetry() {
+    return prepare(prepareInfo, true).thenCompose(this::executePrepared);
+  }
+
+  private ResponseAndTraceId buildResponse(Result result) {
+    ResponseAndTraceId responseAndTraceId = new ResponseAndTraceId();
+    responseAndTraceId.setTracingId(result.getTracingId());
+    Response.Builder responseBuilder = makeResponseBuilder(result);
+    switch (result.kind) {
+      case Void:
+        break;
+      case SchemaChange:
+        // TODO make this non-blocking?
+        persistence.waitForSchemaAgreement();
+        break;
+      case Rows:
+        Payload values = query.getValues();
+        PayloadHandler handler = PayloadHandlers.get(values.getType());
+        try {
+          responseBuilder.setResultSet(
+              Payload.newBuilder()
+                  .setType(query.getValues().getType())
+                  .setData(handler.processResult((Result.Rows) result, query.getParameters())));
+        } catch (Exception e) {
+          throw new CompletionException(e);
+        }
+        break;
+      case SetKeyspace:
+        throw new CompletionException(
+            Status.INVALID_ARGUMENT.withDescription("USE <keyspace> not supported").asException());
+      default:
+        throw new CompletionException(
+            Status.INTERNAL.withDescription("Unhandled result kind").asException());
+    }
+    responseAndTraceId.setResponseBuilder(responseBuilder);
+    return responseAndTraceId;
+  }
+
+  private CompletionStage<Response> executeTracingQueryIfNeeded(
+      ResponseAndTraceId responseAndTraceId) {
+    Response.Builder responseBuilder = responseAndTraceId.responseBuilder;
+    return (responseAndTraceId.tracingIdIsEmpty())
+        ? CompletableFuture.completedFuture(responseBuilder.build())
+        : new QueryTracingFetcher(responseAndTraceId.tracingId, connection, getTracingConsistency())
+            .fetch()
+            .handle(
+                (traces, error) -> {
+                  if (error == null) {
+                    responseBuilder.setTraces(
+                        TraceEventsMapper.toTraceEvents(
+                            traces, responseBuilder.getTraces().getId()));
+                  }
+                  // If error != null, ignore and still return the main result with an empty trace
```

Comment:
This is a departure from the original implementation, but it sounds better to me.

Reply:
Do you mean to return a response without tracing? That sounds like a sane idea +1

Reply:
Yes: if the trace query fails, we still have the main result, so don't fail the whole request.

Reply:
Good catch.